### PR TITLE
Move CMP to `@guardian/libs`

### PR DESCRIPTION
**Pull request by vlkerag**
## What are you changing?

Adds the exports of `@guardian/consent-management-platform` to `@guardian/libs`.

_The original code is marked as being moved here in https://github.com/guardian/consent-management-platform/pull/960._

## Why?

We want to add consent-based purpose checks to exports of `@guardian/libs` that write to user's devices (cookies, storage etc).

Because the CMP already depends on `libs`, consuming the CMP from `libs` would create [a circular dependency](https://dev.to/successgilli/circular-dependency-in-nodejs-and-nestjs-3e1d), which makes it impossible to manage updates etc.

Therefore, we need to co-locate this code.

Refs [#1326](https://github.com/guardian/csnx/issues/1326).

> [!NOTE]
>
> The following changes to the CMP code were required:
>
> - Remove extensions from CMP import specifiers
> - Imports (including mocks) from `@guardian/libs` were updated to local paths, e.g. 
>   ```ts
>   import { log } from '../logger/logger';
>   
>   jest.mock("../storage/storage", () => ({
>     // ...
>   }));
>   ```
> - Types renamed (including exports):
>   - `Callback` → `OnConsentChangeCallback`
>   - `Framework` → `ConsentFramework`

### CODE REVIEW

File: libs/@guardian/libs/cypress/support/e2e.js [hunk 1]
```diff
@@ -14,4 +14,4 @@
 // ***********************************************************
 
 // Import commands.js using ES2015 syntax:
-import './commands';
+import '../e2e/consent-management-platform/support/commands';
```
Comment (oliverabrahams) [inline libs/@guardian/libs/cypress/support/e2e.js:17]: Do we want to have e2e.js and comands.js in separate directories

**Review by oliverabrahams — APPROVED** [review]
Looks good to me!

**Review by akinsola-guardian — APPROVED** [review]
Thanks for this. This looks good.

File: libs/@guardian/libs/src/index.ts [hunk 2]
```diff
@@ -58,3 +58,21 @@ export { storage } from './storage/storage';
 
 export type { Switches } from './switches/@types/Switches';
 export { getSwitches } from './switches/getSwitches';
+
+export type {
+	Callback,
```
Comment (sndrs) [inline libs/@guardian/libs/src/index.ts:63]: i think this name is a bit too generic for a library with many other exports - how about something like `OnConsentChangeCallback`?

File: libs/@guardian/libs/src/index.ts [hunk 3]
```diff
@@ -58,3 +58,21 @@ export { storage } from './storage/storage';
...
+	CMP,
+	ConsentState,
+	Framework,
```
Comment (sndrs) [inline libs/@guardian/libs/src/index.ts:66]: this too is too generic - how about `ConsentFramework`?

**Review by sndrs — CHANGES_REQUESTED** [review]
we need to add a lint to https://github.com/guardian/consent-management-platform for the merge commit.

i'm on it, just hitting the red button to hold off merging till then

File: libs/@guardian/libs/src/index.ts [hunk 4]
```diff
@@ -58,3 +58,21 @@ export { storage } from './storage/storage';
 
 export type { Switches } from './switches/@types/Switches';
 export { getSwitches } from './switches/getSwitches';
+
+export type {
+	Callback as OnConsentChangeCallback,
```
[the remark this answers is not in the document]
Comment (sndrs) [reply libs/@guardian/libs/src/index.ts:63]: yeah agreed - but given you'll need to import from `libs` now, should that means a major version of `libs`?

[the remark this answers is not in the document]
Comment (mxdvl) [reply libs/@guardian/libs/src/index.ts:63]: Oh yes, true! Carry on…

**Review by sndrs — APPROVED** [review]

**Comment by sndrs** [comment]
will merge on monday morning, alongside https://github.com/guardian/consent-management-platform/pull/960

**Pull request merged**